src/lib.rs

```diff
@@ -667,14 +632,172 @@ impl BumpFeeTxBuilder {
+        let path = path.derivation_path_mutex.lock().unwrap().deref().clone();
+        let descriptor_x_key = match descriptor_secret_key.deref() {
+            BdkDescriptorSecretKey::XPrv(descriptor_x_key) => Some(descriptor_x_key),
+            _ => None,
+        }
+        .unwrap();
+        let derived_xprv = descriptor_x_key.xkey.derive_priv(&secp, &path)?;
+        let key_source = match descriptor_x_key.origin.clone() {
+            Some((fingerprint, origin_path)) => (fingerprint, origin_path.extend(path)),
+            None => (descriptor_x_key.xkey.fingerprint(&secp), path),
+        };
+        let derived_descriptor_secret_key = BdkDescriptorSecretKey::XPrv(DescriptorXKey {
+            origin: Some(key_source),
+            xkey: derived_xprv,
+            derivation_path: BdkDerivationPath::default(),
+            wildcard: descriptor_x_key.wildcard,
+        });
+        Ok(Arc::new(Self {
+            descriptor_secret_key_mutex: Mutex::new(derived_descriptor_secret_key),
+        }))
+    }
+
+    fn extend(&self, path: Arc<DerivationPath>) -> Arc<Self> {
+        let descriptor_secret_key = self.descriptor_secret_key_mutex.lock().unwrap();
+        let path = path.derivation_path_mutex.lock().unwrap().deref().clone();
+        let descriptor_x_key = match descriptor_secret_key.deref() {
+            BdkDescriptorSecretKey::XPrv(descriptor_x_key) => Some(descriptor_x_key),
+            _ => None,
+        }
+        .unwrap();
+        let extended_path = descriptor_x_key.derivation_path.extend(path);
+        let extended_descriptor_secret_key = BdkDescriptorSecretKey::XPrv(DescriptorXKey {
+            origin: descriptor_x_key.origin.clone(),
+            xkey: descriptor_x_key.xkey,
+            derivation_path: extended_path,
+            wildcard: descriptor_x_key.wildcard,
+        });
+        Arc::new(Self {
+            descriptor_secret_key_mutex: Mutex::new(extended_descriptor_secret_key),
+        })
+    }
+
+    fn as_public(&self) -> Arc<DescriptorPublicKey> {
+        let secp = Secp256k1::new();
+        let descriptor_public_key = self
+            .descriptor_secret_key_mutex
+            .lock()
+            .unwrap()
+            .as_public(&secp)
+            .unwrap();
+        Arc::new(DescriptorPublicKey {
+            descriptor_public_key_mutex: Mutex::new(descriptor_public_key),
+        })
+    }
+
+    fn as_string(&self) -> String {
+        self.descriptor_secret_key_mutex.lock().unwrap().to_string()
+    }
+}
+
+struct DescriptorPublicKey {
+    descriptor_public_key_mutex: Mutex<BdkDescriptorPublicKey>,
+}
+
+impl DescriptorPublicKey {
+    fn derive(&self, path: Arc<DerivationPath>) -> Result<Arc<Self>, BdkError> {
+        let secp = Secp256k1::new();
+        let descriptor_public_key = self.descriptor_public_key_mutex.lock().unwrap();
+        let path = path.derivation_path_mutex.lock().unwrap().deref().clone();
+        let descriptor_x_key = match descriptor_public_key.deref() {
```

Comment:
I think this match code and other's like it in your code would be clearer like this:
```rust
match descriptor_secret_key.deref() {
      BdkDescriptorSecretKey::XPrv(descriptor_x_key) => {
          let derived_xprv = descriptor_x_key.xkey.derive_priv(&secp, &path)?;
          let key_source = match descriptor_x_key.origin.clone() {
              Some((fingerprint, origin_path)) => (fingerprint, origin_path.extend(path)),
              None => (descriptor_x_key.xkey.fingerprint(&secp), path),
          };
          let derived_descriptor_secret_key = BdkDescriptorSecretKey::XPrv(DescriptorXKey {
              origin: Some(key_source),
              xkey: derived_xprv,
              derivation_path: BdkDerivationPath::default(),
              wildcard: descriptor_x_key.wildcard,
          });
          Ok(Arc::new(Self {
              descriptor_secret_key_mutex: Mutex::new(derived_descriptor_secret_key),
          }))
      },
      BdkDescriptorSecretKey::SinglePriv(_) => {
          unreachable!()
      }
  }
```
This way it's clear that since we're not supporting single keys the alternative branch is unreachable (which will still cause a panic if for some reason it is reached). 

At some point we should support single keys.. and then in the derive/extend functions we'll need to throw a run-time Err, which is less than ideal. I'd rather avoid that, but with no support for generics in the APIs we can build with uniffi it may be the least ugly approach. But not a concern for this PR.